Kernel/Term.hpp

```diff
@@ -903,8 +913,11 @@ class alignas(8) Term
   unsigned _hasInterpretedConstants : 1;
   /** If true, the object is an equality literal between two variables */
   unsigned _isTwoVarEquality : 1;
-  /** Weight of the symbol */
+  /** Weight of the symbol, i.e. sum of symbol and variable occurrences. */
   unsigned _weight;
+  /** Cached weight of the term for KBO, otherwise -1 and invalid. Note that
+   * KBO symbol weights are not necessarily 1, so this can differ from @b _weight. */
+  int _kboWeight;
```

Comment:
So `_kboWeight` is pre-computed on a per-term version. But it's actually not only dependent on the term but also on the KBO instance that computes the weight (i.e. it depends on which weight function that KBO uses). Thus having one value per term implicitly assumes that we always only have exactly one KBO instances, which is an assumption that could be broken in the future. I'm thinking for example multiple vampire instances running on one machine using the same memory for terms & so but using different weight functions, etc. Sooo it would maybe be good to add an assertion here that makes sure that we are only using one KBO instance to set the weight and only set the weight once.

I'm thinking adding a field `void* _kboInstance` that is initialized with nullpoint and needs to be set with `void setKboWeight(int weight, void* kbo)` (using `void*` so noone has the idea of calling that kbo instance), and when we call `setKboWeight` we add `ASS(_kboInstance == nullptr)`, and we also change `getKboWeight()` to `getKboWeight(void* kbo)` which checks `_kboInsatnce == kbo` before returning the weight. What are you thinking about that?
(obviously hiding all that safety code behing `#if VDEBUG`)

Reply:
Another case where we could have multiple KBO instances is unit testing ofc.

Reply:
Good point! 

But note that we already cache orientations of equalities within the literals and I think there, there is some code that looks into env and only if the ordering we are computing the orientation with is the same as some main one, are we allowed to use / fill the cache. Couldn't that be a solution here as well?

Reply:
Good shout! That's the very same thing and I'd also thought about using `env.ordering`
I'd still opt for having an additional `Ordering*` parameter as that insures two things: People will always need to have access to the ordering wrt when they call the respective function (thus think about that things are ordering-dependent), and also it makes things more future-proof: If we decide we do want the option to have multiple orderings supported at once, we could have a `static DHMap<...> Term::_kboWeights` behind `setKboWeight` and so. 
But I don't think this is strictly necessary, it's just how i'd do it. I'm fine with either as long as there is some assertion that catches the potential future bugs :)